### PR TITLE
embed excel interop types

### PR DESCRIPTION
**Pull request by pinzart90**
Changes:
1. There are now 3 DSOffice projects in the source code:
  For net48 I kept the DSOffice as it was (if we cut 2.19 before this is merged I can unify the solutions)
  For net6 I split the old DSOffice project into 2 new Projects: DSOfficeNodes and DSOfficeUtils
  DSOfficeNodes => Contains almost all of the Public API (that will be imported as nodes in Dynamo) and image/customization resources
  DSOfficeUtilities => Contains all the internal APIs and embeds the interop excel dependencies and string resources
  By having a satellite assembly (DSOfficeUtilities) to hold all the embedded Excel interop types, we avoid having to import those same types in Dynamo.
2. I kept the ProtoCore changes (to handle System.Missing types). I can add tests for these if we decide it is useful.

Notes:
Net48 will not have any of these changes. It will have

### CODE REVIEW

**Comment by mjkkirschner** [comment]
```
The task "ResolveComReference" is not supported on the .NET Core version of MSBuild. Please use the .NET Framework version of MSBuild. See https://aka.ms/msbuild/MSB4803 for further details. [/home/runner/work/Dynamo/Dynamo/Dynamo/src/Libraries/DSOffice/DSOffice.csproj]
    559 Warning(s)
    1 Error(s)
```

bummer

**Comment by pinzart90** [comment]
> bummer

COM references only work on windows. So we can keep it for bot All and Core solutions but only Windows
I do not remember if there was any value for Linux..

**Comment by mjkkirschner** [comment]
> > bummer
> 
> COM references only work on windows. So we can keep it for bot All and Core solutions but only Windows I do not remember if there was any value for Linux..

so, I assume this means the excel nodes never had a chance of working on linux ... given no linux excel, not even considering COM?

**Comment by pinzart90** [comment]
> n no linux excel,

Yes, there is no excel on linux (AFAIK). We could remove COM from the linux build and just deliver all assemblies (office.dll, excel.interop.dll etc)

File: src/Libraries/DSOffice/DSOffice.csproj [hunk 1]
```diff
@@ -13,17 +13,33 @@
...
-    </ItemGroup>
-    <ItemGroup Condition=" '$(TargetFramework)' == 'net6.0' ">
-        <PackageReference Include="Microsoft.Office.Interop.Excel" Version="15.0.4795.1001" />
-        <PackageReference Include="MicrosoftOfficeCore" Version="15.0.0" />
-    </ItemGroup>
-  <ItemGroup Label ="Common deps">
+  <ItemGroup Condition=" '$(TargetFramework)' == 'net48' ">
+    <PackageReference Include="Microsoft.Office.Interop.Excel" Version="15.0.4795.1001">
+        <ExcludeAssets>runtime</ExcludeAssets>
+    </PackageReference>
```
Comment (pinzart90) [inline src/Libraries/DSOffice/DSOffice.csproj:19]: net48 add excel as a reference but do not deliver it with Dynamo (it will be found in the GAC at runtime)

File: src/Libraries/DSOffice/DSOffice.csproj [hunk 2]
```diff
@@ -13,17 +13,33 @@
...
+    <Reference Include="WindowsBase" />
+  </ItemGroup>
+  <ItemGroup Condition=" '$(TargetFramework)' == 'net6.0' ">
+    <PackageReference Include="Microsoft.Office.Interop.Excel" Version="15.0.4795.1001">
+        <ExcludeAssets>all</ExcludeAssets>
```
Comment (pinzart90) [inline src/Libraries/DSOffice/DSOffice.csproj:24]: Do no add as typical reference, just use the path of the package to add the dll as an embedded reference

File: src/DynamoCore.net6.sln [hunk 3]
```diff
@@ -372,11 +372,9 @@ Global
 		{6CD0F0CF-8199-49F9-B0EA-0B9598B44419}.Release|NET60_Windows.ActiveCfg = Release|NET60_Windows
 		{6CD0F0CF-8199-49F9-B0EA-0B9598B44419}.Release|NET60_Windows.Build.0 = Release|NET60_Windows
 		{9B4FDC96-E2F9-4B8F-894A-4294405D50E7}.Debug|NET60_Linux.ActiveCfg = Debug|NET60_Linux
-		{9B4FDC96-E2F9-4B8F-894A-4294405D50E7}.Debug|NET60_Linux.Build.0 = Debug|NET60_Linux
 		{9B4FDC96-E2F9-4B8F-894A-4294405D50E7}.Debug|NET60_Windows.ActiveCfg = Debug|NET60_Windows
 		{9B4FDC96-E2F9-4B8F-894A-4294405D50E7}.Debug|NET60_Windows.Build.0 = Debug|NET60_Windows
 		{9B4FDC96-E2F9-4B8F-894A-4294405D50E7}.Release|NET60_Linux.ActiveCfg = Release|NET60_Linux
-		{9B4FDC96-E2F9-4B8F-894A-4294405D50E7}.Release|NET60_Linux.Build.0 = Release|NET60_Linux
```
[the remark this answers is not in the document]
Comment (pinzart90) [reply src/DynamoCore.net6.sln:379]: There are 2 APIs that would work on Linux:
ImportCSV and ExportCSV
Not sure if it is worth the work at the moment (we could add it later if it is needed o linux)

[the remark this answers is not in the document]
Comment (pinzart90) [reply src/DynamoCore.net6.sln:379]: not really, If we want them on Linux I can conditionally compile only those APIs (maybe move them to a separate .cs to make it easier)

File: src/Engine/ProtoCore/Parser/AssociativeAST.cs [hunk 4]
```diff
@@ -2803,18 +2804,45 @@ public static BooleanNode BuildBooleanNode(bool value)
...
+
+        /// <summary>
+        /// Builds an AssociativeNode (int, double, null etc) from the input value or type
+        /// </summary>
+        /// <param name="value">The value of the object</param>
+        /// <param name="type">The target type that the "value" parameter represents</param>
+        /// <returns>AssociativeNode</returns>
+        internal static AssociativeNode BuildPrimitiveNodeFromObjectAndType(object value, System.Type type)
+        {
+            if (value is Missing)
```
[the remark this answers is not in the document]
Comment (aparajit-pratap) [reply src/Engine/ProtoCore/Parser/AssociativeAST.cs:2827]: Why do we want to import these interop types? Which nodes expose these types?

File: src/DynamoCore.net6.sln (hunk 3 above)
[the remark this answers is not in the document]
Comment (aparajit-pratap) [reply src/DynamoCore.net6.sln:379]: I meant if you're removing the project here for Linux, you'll need a new project, won't you?

File: src/Libraries/DSOffice/DSOffice.csproj (hunk 2 above)
Comment (aparajit-pratap) [reply src/Libraries/DSOffice/DSOffice.csproj:24]: Can you explain again why you're adding it as an embedded reference?

**Review by aparajit-pratap — COMMENTED** [review]
Can you verify that no additional interop types are imported into the VM?

File: src/Engine/ProtoCore/Parser/AssociativeAST.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (pinzart90) [reply src/Engine/ProtoCore/Parser/AssociativeAST.cs:2827]: Well ...all the Excel interop types used by the existing nodes are imported (since the types are embedded they appear as any other types in the assembly).
image

[the remark this answers is not in the document]
Comment (pinzart90) [reply src/Engine/ProtoCore/Parser/AssociativeAST.cs:2827]: All these types do have a COM specific atribute {[System.Runtime.InteropServices.ComImportAttribute()]}
We could try to filter out COM types that are coming from DSOffice.dll

Dynamo does support embedded types. There might be valid cases where users want to import nodes from embedded types. We had this issue reported https://forum.dynamobim.com/t/fail-to-load-two-packages-access-to-the-same-api/83556/30
This was added to our docs as a response

File: src/Libraries/DSOffice/DSOffice.csproj (hunk 2 above)
Comment (pinzart90) [reply src/Libraries/DSOffice/DSOffice.csproj:24]: I chose embedded types so that we can avoid having to ship the Interop.Excel.dll and office.dll (these were in the GAC for net48)
They might cause conflicts with host integrators

File: src/Dynamo.All.Net6.sln [hunk 5]
```diff
@@ -194,311 +195,619 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "DynamoServicesTests", "..\t
...
+		{6DD7D985-9057-4678-A2F3-E7A2531F31F4}.Release|NET60_Linux.ActiveCfg = Release|NET60_Linux
+		{6DD7D985-9057-4678-A2F3-E7A2531F31F4}.Release|NET60_Linux.Build.0 = Release|NET60_Linux
 		{6DD7D985-9057-4678-A2F3-E7A2531F31F4}.Release|NET60_Windows.ActiveCfg = Release|NET60_Windows
 		{6DD7D985-9057-4678-A2F3-E7A2531F31F4}.Release|NET60_Windows.Build.0 = Release|NET60_Windows
+		{6CC1E399-B5A5-4FA2-A98C-5D8B2BA280A1}.Debug|NET60_Linux.ActiveCfg = Debug|NET60_Linux
+		{6CC1E399-B5A5-4FA2-A98C-5D8B2BA280A1}.Debug|NET60_Windows.ActiveCfg = Debug|NET60_Windows
+		{6CC1E399-B5A5-4FA2-A98C-5D8B2BA280A1}.Debug|NET60_Windows.Build.0 = Debug|NET60_Windows
+		{6CC1E399-B5A5-4FA2-A98C-5D8B2BA280A1}.Release|NET60_Linux.ActiveCfg = Release|NET60_Linux
+		{6CC1E399-B5A5-4FA2-A98C-5D8B2BA280A1}.Release|NET60_Windows.ActiveCfg = Release|NET60_Windows
+		{6CC1E399-B5A5-4FA2-A98C-5D8B2BA280A1}.Release|NET60_Windows.Build.0 = Release|NET60_Windows
```
[the remark this answers is not in the document]
Comment (pinzart90) [reply src/Dynamo.All.Net6.sln:810]: Yes, thanks. Will fix once we decide on the other parts. I want to first figure out how to best organize the new projects and the outputs (and on what target framework)

**Comment by pinzart90** [comment]
> Can you verify that no additional interop types are imported into the VM?

I will add a test for it

**Comment by pinzart90** [comment]
I still have to add a couple of tests

File: test/DynamoCoreTests/UnitTestBase.cs [hunk 6]
```diff
@@ -65,10 +65,6 @@ public static string TestDirectory
         public virtual void Setup()
         {
             SetupDirectories();
-#if NETFRAMEWORK
-            
-            DSOffice.ExcelInterop.ShowOnStartup = false;
-#endif
```
Comment (pinzart90) [inline test/DynamoCoreTests/UnitTestBase.cs:71]: moved this to DynamoMSOfficeTests Setup.cs

File: test/Engine/FFITarget/ClassFunctionality.cs [hunk 7]
```diff
@@ -179,4 +180,13 @@ public override string ToString()
             throw new NotImplementedException();
         }
     }
+
+    public class MissingClass
+    {
+        public static long MissingMethod([In, Optional] object value)
```
Comment (pinzart90) [inline test/Engine/FFITarget/ClassFunctionality.cs:186]: This mimics some existing types from Excel.Interop
https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/named-and-optional-arguments

File: src/Libraries/DSOffice/DSOffice.csproj [hunk 8]
```diff
@@ -13,18 +13,12 @@
...
         </PackageReference>
         <Reference Include="WindowsBase" />
-    </ItemGroup>
-    <ItemGroup Condition=" '$(TargetFramework)' == 'net6.0' ">
-        <PackageReference Include="Microsoft.Office.Interop.Excel" Version="15.0.4795.1001" />
-        <PackageReference Include="MicrosoftOfficeCore" Version="15.0.0" />
-    </ItemGroup>
-  <ItemGroup Label ="Common deps">
-	<PackageReference Include="DocumentFormat.OpenXml" Version="2.12.3" CopyXML="true" />
+        <PackageReference Include="DocumentFormat.OpenXml" Version="2.12.3" CopyXML="true" />
```
Comment (pinzart90) [inline src/Libraries/DSOffice/DSOffice.csproj:21]: this project will only be compiled on net48

File: src/Libraries/DSOfficeNodes/Excel.cs [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+        /// <param name="overWrite"> Toggle to clear spreadsheet before writing</param>
+        /// <param name="writeAsString">Toggle to switch between writing cell values as strings</param>
+        /// <search>office,excel,spreadsheet</search>
+        /// <returns>Boolean indicating if writing to spreadsheet is successful.</returns>
+        public static bool OpenXMLExportExcel(string filePath, string sheetName, object[][] data, int startRow = 0, int startColumn = 0, bool overWrite = false, bool writeAsString = false)
+        {
+            return OpenXmlHelper.Write(filePath, sheetName, data, startRow, startColumn, overWrite, writeAsString);
+        }
+    }
+}
```
Comment (pinzart90) [inline src/Libraries/DSOfficeNodes/Excel.cs:293]: THese are the public classes and methods from the original DSOffice

File: src/Libraries/DSOfficeNodes/DSOffice.csproj [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        <SubType>Designer</SubType>
+        <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </Content>
+    <Content Include="DSOffice_DynamoCustomization.xml">
+        <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </Content>
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="CSVNodes.cs" />
+    <Compile Include="Excel.cs" Condition=" '$(RuntimeIdentifier)' == 'win10-x64' " />
```
Comment (pinzart90) [inline src/Libraries/DSOfficeNodes/DSOffice.csproj:27]: Excel.cs will only be compiled on windows

File: src/Libraries/DSOfficeNodes/Properties/AssemblyInfo.cs [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+
+// General Information about an assembly is controlled through the following 
+// set of attributes. Change these attribute values to modify the information
+// associated with an assembly.
+[assembly: AssemblyTitle("DSOffice")]
+[assembly: AssemblyCulture("")]
+
+[assembly: InternalsVisibleTo("DynamoMSOfficeTests")]
+[assembly: TypeForwardedTo(typeof(DSOffice.WorkBook))]
+[assembly: TypeForwardedTo(typeof(DSOffice.WorkSheet))]
```
Comment (pinzart90) [inline src/Libraries/DSOfficeNodes/Properties/AssemblyInfo.cs:12]: I moved these types to the OfficeUtilities assembly because they are used there and they are not actually visible in Dynamo

File: src/Libraries/DSOfficeUtilities/DSOfficeUtilities.csproj [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    <EmbeddedResource Update="Properties\Resources.resx">
+      <Generator>ResXFileCodeGenerator</Generator>
+      <LastGenOutput>Resources.Designer.cs</LastGenOutput>
+    </EmbeddedResource>
+    <EmbeddedResource Update="Properties\Resources.en-US.resx">
+      <Generator>ResXFileCodeGenerator</Generator>
+      <LastGenOutput>Resources.en-US.Designer.cs</LastGenOutput>
+    </EmbeddedResource>
+  </ItemGroup>
+</Project>
```
Comment (pinzart90) [inline src/Libraries/DSOfficeUtilities/DSOfficeUtilities.csproj:77]: This project has all the internal APIs of the original DSOffice project and the WorkSheet and WorkBook public types.

File: src/Libraries/DSOfficeNodes/CSVNodes.cs [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+    public static partial class Data
+    {
+        /// <summary>
+        ///     Write a list of lists into a file using a comma-separated values 
+        ///     format. Outer list represents rows, inner lists represent columns. 
+        /// </summary>
+        /// <param name="filePath">Path to write to</param>
+        /// <param name="data">List of lists to write into CSV</param>
+        /// <search>write,text,file</search>
+        public static void ExportCSV(string filePath, object[][] data)
```
Comment (mjkkirschner) [inline src/Libraries/DSOfficeNodes/CSVNodes.cs:21]: so I think this is a problem. You have removed the original assembly so you can no longer use type forwarding to redirect callers to the new assembly location.

This will likely break python and c# usage of these public methods and it may also break graph usage. What about continuing to build the DSOffice project as facade only even for net6?

Also please verify graphs containing these nodes open correctly after the reorganization.

File: src/Tools/Md2Html/Md2Html.csproj [hunk 14]
```diff
@@ -41,7 +41,8 @@
     <Target Name="Publish net6" Condition=" '$(TargetFramework)' == 'net6.0' And '$([MSBuild]::IsOSPlatform(`Windows`))' == 'True' " AfterTargets="Build">
         <CallTarget Targets="Publish" />
         <Exec Command="if not exist $(SharedOutputPath) mkdir $(SharedOutputPath)"/>
-        <Copy SourceFiles="$(OutputPath)win10-x64\publish\Md2Html.exe" DestinationFolder="$(SharedOutputPath)" />
+        <Copy Condition=" '$(RuntimeIdentifier)' == 'win10-x64' " SourceFiles="$(OutputPath)$(RuntimeIdentifier)\publish\Md2Html.exe" DestinationFolder="$(SharedOutputPath)" />
```
Comment (mjkkirschner) [inline src/Tools/Md2Html/Md2Html.csproj:44]: unrelated changes?

**Comment by mjkkirschner** [comment]
looks good except I am concerned about the public breaks with no type forwarding - even for 3.0, if we can keep graphs / code working we should.

File: test/Engine/FFITarget/ClassFunctionality.cs (hunk 7 above)
Comment (mjkkirschner) [reply test/Engine/FFITarget/ClassFunctionality.cs:186]: I read that document, but didn't see any syntax like this `[In, Optional]` - what should I be searching for?

File: src/Libraries/DSOfficeNodes/CSVNodes.cs (hunk 13 above)
Comment (pinzart90) [reply src/Libraries/DSOfficeNodes/CSVNodes.cs:21]: But I kept the same output name. 
DSOfficeNodes/DSOffice.csproj => will produce DSOffice.dll
I hoped that would allow typeforwarding

File: test/Engine/FFITarget/ClassFunctionality.cs (hunk 7 above)
Comment (pinzart90) [reply test/Engine/FFITarget/ClassFunctionality.cs:186]: This section
```
You can also declare optional parameters by using the .NET [OptionalAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.optionalattribute) class. OptionalAttribute parameters do not require a default value.
```

**Review by mjkkirschner — APPROVED** [review]
LGTM - but I think merging after 2.19 cut is fine to avoid the headache.

File: src/Libraries/DSOfficeNodes/CSVNodes.cs (hunk 13 above)
Comment (pinzart90) [reply src/Libraries/DSOfficeNodes/CSVNodes.cs:21]: Yes, the same fully qualified name

File: src/Libraries/DSOfficeUtilities/Excel.cs [hunk 15]
```diff
@@ -0,0 +1,533 @@
...
+        internal static object[][] WriteData(string filePath, string sheetName, int startRow, int startCol, object[][] data, bool overWrite = false, bool writeAsString = false)
+        {
+            WorkBook wb = new WorkBook(filePath);
+            WorkSheet ws = new WorkSheet(wb, sheetName, overWrite);
+            ws = ws.WriteData(startRow, startCol, data, writeAsString);
+            return ws.Data;
+        }
+    }
+
+    [IsVisibleInDynamoLibrary(false)]
```
Comment (aparajit-pratap) [inline src/Libraries/DSOfficeUtilities/Excel.cs:169]: I think we can remove these attributes as this project is no longer a node library.

**Review by aparajit-pratap — COMMENTED** [review]
Does importing the same DLL name [here](https://github.com/DynamoDS/Dynamo/blob/29e422725c4c95984445c0d3236eae30a9a693fe/src/DynamoApplications/PathResolvers.cs#L31) just work? Do we need to change anything in this list?

File: src/Libraries/DSOfficeUtilities/OpenXmlHelper.cs [hunk 16]
```diff
@@ -0,0 +1,656 @@
...
+            }
+            else
+            {
+                // This is for long and double. Also acts as the default behavior for things we do not specially handle.
+                cell.CellValue = new CellValue(Convert.ToString(value, CultureInfo.InvariantCulture));
+                cell.DataType = null;
+            }
+        }
+    }
+}
```
Comment (aparajit-pratap) [inline src/Libraries/DSOfficeUtilities/OpenXmlHelper.cs:656]: So all of this code doesn't duplicate what's already there in `DSOffice` as `DSOffice `isn't included the .net6 sln and the `DSOfficeUtilities` is not included in the net48 sln correct?

**Comment by pinzart90** [comment]
> Does importing the same DLL name [here](https://github.com/DynamoDS/Dynamo/blob/29e422725c4c95984445c0d3236eae30a9a693fe/src/DynamoApplications/PathResolvers.cs#L31) just work? Do we need to change anything in this list?
It just works because the dependency on DSOfficeUtilities is resolved by dotnet or one of Dynamo's assembly resolvers.

**Review by aparajit-pratap — APPROVED** [review]
Other than the cleanup of the unnecessary "isvisible" attributes in DSOfficeUtilities, LGTM.

**Comment by pinzart90** [comment]
Tests passed here https://master-5.jenkins.autodesk.com/job/Dynamo/job/DynamoSelfServe/job/pullRequestValidation/10602/
